### PR TITLE
SM Model Jupyter Notebook Tutorial

### PR DESCRIPTION
**Pull request by tuzhucheng**
Adding a notebook tutorial stepping through the SM CNN as we discussed before for future reference.

See a preview of the notebook at https://nbviewer.jupyter.org/github/tuzhucheng/Castor/blob/sm-model-tutorial/sm_cnn/tutorial.ipynb

@rosequ do you mind taking a look?

### CODE REVIEW

**Comment by rosequ** [comment]
@tuzhucheng having a look at it. Could you also upload it to the repo, directly? You can view them directly on GitHub.

**Comment by rosequ** [comment]
Also, I think it would be easier for me to comment.

**Comment by tuzhucheng** [comment]
Here is the repo link: https://github.com/tuzhucheng/Castor/blob/sm-model-tutorial/sm_cnn/tutorial.ipynb

**Comment by rosequ** [comment]
- :param no_ext_feats: whether or not to use additional features X_feat. Change it to "no additional external features"
- A few lines on the nn_sequential---the structure, activation function, forward() call---would help the reader, I think
- Also, would it be possible to add a small tutorial on the bridge file, this would help one know how load and use a pre-trained model
- Explain what Linear, dropout() and hidden() does.
It looks good otherwise. I so wish I could comment on the notebook directly!

**Comment by lintool** [comment]
This doesn't have bilinear sim component, right? Point that out and references this: 

+ Jinfeng Rao, Hua He, and Jimmy Lin. Experiments with Convolutional Neural Network Models for Answer Selection. Proceedings of the 40th Annual International ACM SIGIR Conference on Research and Development in Information Retrieval (SIGIR 2017), August 2017, Tokyo, Japan.

Can we find a better example question? I don't like gangs...

**Comment by lintool** [comment]
BTW, this is great! Thanks for working on this.

I don't think showing all dimensions of the word embeddings vectors is particularly helpful. Maybe just show a snippet, and most importantly, the dimensions?

**Comment by tuzhucheng** [comment]
Addressed all comments except for the one about the bridge file. @rosequ isn't the main.py already capable of loading and using pre-trained model, i.e. `trained_model = QAModel.load('../../models/sm_model/sm_model.TrecQA.TRAIN-ALL.2017-04-02.castor')`?

**Comment by rosequ** [comment]
@tuzhucheng yes I saw it. I thought some of the code from `main` isn't necessary, right? Like this one:
```
trained_model_weights = list(trained_model.parameters())
evaluator = Trainer(trained_model, 0.001, 0.0, False, vec_dim)

evaluator.load_input_data('../../data/TrecQA', cache_file, None, None, 'raw-dev')

questions, sentences, labels, maxlen_q, maxlen_s, ext_feats = evaluator.data_splits['raw-dev']
```

Just the structure for loading and passing the right parameters would be helpful, I guess.

**Comment by rosequ** [comment]
Looks good to me!
